crates/cext/src/circuit.rs

```diff
@@ -883,24 +888,36 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
         panic!("Invalid index")
     }
     let packed_inst = &circuit.data()[index];
-    let qargs = circuit.get_qargs(packed_inst.qubits);
-    let mut qargs_vec: Vec<u32> = qargs.iter().map(|x| x.0).collect();
-    let cargs = circuit.get_cargs(packed_inst.clbits);
-    let mut cargs_vec: Vec<u32> = cargs.iter().map(|x| x.0).collect();
-    let params = packed_inst.params_view();
-    let mut params_vec: Vec<f64> = params
-        .iter()
-        .map(|x| match x {
-            Param::Float(val) => *val,
-            _ => unreachable!("Invalid parameter on instruction"),
-        })
-        .collect();
-    let out_qargs = qargs_vec.as_mut_ptr();
-    std::mem::forget(qargs_vec);
-    let out_cargs = cargs_vec.as_mut_ptr();
-    std::mem::forget(cargs_vec);
-    let out_params = params_vec.as_mut_ptr();
-    std::mem::forget(params_vec);
+    let mut qargs = {
+        let qargs = circuit.get_qargs(packed_inst.qubits);
+        let qargs_vec: Vec<u32> = qargs.iter().map(|x| x.0).collect();
+        qargs_vec.into_boxed_slice()
+    };
+    let mut cargs = {
+        let cargs = circuit.get_cargs(packed_inst.clbits);
+        let cargs_vec: Vec<u32> = cargs.iter().map(|x| x.0).collect();
+        cargs_vec.into_boxed_slice()
+    };
+    let mut params = {
+        let params = packed_inst.params_view();
+        let params_vec: Vec<f64> = params
+            .iter()
+            .map(|x| match x {
+                Param::Float(val) => *val,
+                _ => panic!("Invalid parameter on instruction"),
+            })
+            .collect();
+        params_vec.into_boxed_slice()
+    };
+    let out_qargs = qargs.as_mut_ptr();
+    let out_qargs_len = qargs.len() as u32;
+    let _ = Box::into_raw(qargs);
```

Comment:
This is subtle enough that I think it might warrant a comment something like:

```suggestion
    // .as_mut_ptr() returns the mutable pointer to the slice/array
    // so that we point to [T] instead of T, Box::into_raw()
    // consumes the `Box` so the allocation will not be freed when
    // it is dropped
    let out_qargs = qargs.as_mut_ptr();
    let out_qargs_len = qargs.len() as u32;
    let _ = Box::into_raw(qargs);
```

Reply:
Done in 38e60a051a11f98a992a7d0948eea16be1b29645